routing/grandcentral/server.go

```diff
@@ -138,3 +100,92 @@ func (s *Server) handleMessage(
 
 var _ proxy.RequestHandler = &Server{}
 var _ proxy.Proxy = &Server{}
+
+func getRoutingRecord(ds datastore.Datastore, k util.Key) (*dhtpb.Record, error) {
+	dskey := k.DsKey()
+	val, err := ds.Get(dskey)
+	if err != nil {
+		return nil, errors.Wrap(err)
+	}
+	recordBytes, ok := val.([]byte)
+	if !ok {
+		return nil, fmt.Errorf("datastore had non byte-slice value for %v", dskey)
+	}
+	var record dhtpb.Record
+	if err := proto.Unmarshal(recordBytes, &record); err != nil {
+		return nil, errors.New("failed to unmarshal dht record from datastore")
+	}
+	return &record, nil
+}
+
+func putRoutingRecord(ds datastore.Datastore, k util.Key, value *dhtpb.Record) error {
+	data, err := proto.Marshal(value)
+	if err != nil {
+		return err
+	}
+	dskey := k.DsKey()
+	// TODO namespace
+	if err := ds.Put(dskey, data); err != nil {
+		return err
+	}
+	return nil
+}
+
+func putRoutingProviders(ds datastore.Datastore, k util.Key, newRecords []*dhtpb.Message_Peer) error {
+	log.Event(context.Background(), "putRoutingProviders", &k)
+	oldRecords, err := getRoutingProviders(ds, k)
+	if err != nil {
+		return err
+	}
+	mergedRecords := make(map[string]*dhtpb.Message_Peer)
+	for _, provider := range oldRecords {
+		mergedRecords[provider.GetId()] = provider // add original records
+	}
+	for _, provider := range newRecords {
+		mergedRecords[provider.GetId()] = provider // overwrite old record if new exists
+	}
+	var protomsg dhtpb.Message
+	protomsg.ProviderPeers = make([]*dhtpb.Message_Peer, 0, len(mergedRecords))
+	for _, provider := range mergedRecords {
+		protomsg.ProviderPeers = append(protomsg.ProviderPeers, provider)
+	}
+	data, err := proto.Marshal(&protomsg)
+	if err != nil {
+		return err
+	}
+	return ds.Put(providerKey(k), data)
+}
+
+func storeProvidersToPeerstore(ps peer.Peerstore, p peer.ID, providers []*dhtpb.Message_Peer) {
+	for _, provider := range providers {
+		providerID := peer.ID(provider.GetId())
+		if providerID != p {
+			log.Errorf("provider message came from third-party %s", p)
```

Comment:
there should be this check as well in `putRoutingProviders` or this should return an error. Otherwise the record _will_ be stored (despite us not collecting their address).